plugins/poll/lib/PollActions.php

```diff
@@ -0,0 +1,256 @@
+	}
+
+	private static function isValidPollIdStructure($id)
+	{
+		self::init();
+		$idElements = explode(self::ID_SEPARATOR_CHAR, $id);
+		if (count($idElements) === self::ID_NUM_ELEMENTS ) {
+			$pollType = $idElements[0];
+			$hash = $idElements[1];
+			$key = $idElements[2];
+			$simulatedHash = hash_hmac('SHA256', self::$kalturaSecret, $key);
+			$isHashOk = strcmp($hash, $simulatedHash) === 0;
+			$validPollType = self::isValidPollType($pollType);
+			return $isHashOk && $validPollType;
+
+		}
+		return false;
+	}
+
+	/* Poll Vote Actions */
+	public static function vote($params)
+	{
+		$pollId = $params['pollId'];
+		$userId = $params['userId'];
+		$ansIds = $params['answerIds'];
+		return self::setVote($pollId, $userId, $ansIds);
+	}
+
+	public static function setVote($pollId, $userId, $ansIds)
+	{
+		if (!$pollId || !$userId || !$ansIds)
+			throw new Exception('Missing parameter for vote action');
+		$answers = explode(self::ANSWER_SEPARATOR_CHAR, $ansIds);
+		if (self::isValidPollIdStructure($pollId)) {
+			// check early user vote
+			$previousAnswers = self::$pollsCacheHandler->setCacheVote($userId, $pollId, $answers);
+			if ($previousAnswers) {
+				self::$pollsCacheHandler->decrementAnsCounter($pollId, $previousAnswers);
+			} else {
+				self::$pollsCacheHandler->incrementPollVotersCount($pollId);
+			}
+			self::$pollsCacheHandler->increaseAnsCounter($pollId, $answers);
+			return "Successfully voted";
+		}
+		return "Failed to vote due to bad poll id structure";
+	}
+
+	/* Poll Get Votes Actions */
+
+	public static function getVotes($pollId, $ansIds)
+	{
+		if (!$pollId || !$ansIds)
+			throw new Exception('Missing parameter for getVotes action');
+		self::init();
+		$answers = explode(self::ANSWER_SEPARATOR_CHAR, $ansIds);
+		$pollVotes = new PollVotes($pollId);
+		$pollVotes->setNumVoters(self::$pollsCacheHandler->getPollVotersCount($pollId));
+		foreach($answers as $ansId) {
+			$answerCount = self::$pollsCacheHandler->getAnswerCounter($pollId, $ansId);
+			$pollVotes->addAnswerCounter($ansId, $answerCount);
+		}
+		return $pollVotes;
+	}
+
+}
+
+class PollCacheHandler
+{
+
+	private $cache;
+
+	private $cacheTTL;
+
+	public function __construct($cacheTTL)
+	{
+		$this->cache = kCacheManager::getSingleLayerCache(kCacheManager::CACHE_TYPE_CACHE_ONLY_ACTIONS);
+		$this->cacheTTL = $cacheTTL;
+	}
+
+	public function setCacheVote($userId, $pollId, $ansIds)
+	{
+		$userVoteKey = self::getPollUserVoteCacheKey($pollId, $userId);
+		if (!$this->cache->add($userVoteKey, $ansIds, $this->cacheTTL)) {
+			$earlyVoteAnsIds = $this->cache->get($userVoteKey);
+			$this->cache->set($userVoteKey, $ansIds);
+			return $earlyVoteAnsIds;
+		}
+		return null;
+	}
+
+	public function getAnswerCounter($pollId, $ansId)
+	{
+		$key = $this->getPollAnswerCounterCacheKey($pollId, $ansId);
+		$counter = $this->cache->get($key);
+		if (!$counter)
+			return 0;
+		return $counter;
+	}
+
+
+	public function increaseAnsCounter($pollId, $ansIds)

```

Comment:
Name is not clear. Is it for one answer? many answers? rename.
Also , increment is a more clearer terminology , rename.